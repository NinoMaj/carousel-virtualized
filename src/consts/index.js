export const RESIZE_THROTTLE_THRESHOLD = 500;
