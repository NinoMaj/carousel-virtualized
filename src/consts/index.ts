// tslint:disable-next-line:export-name
export const RESIZE_THROTTLE_THRESHOLD: number = 500;
