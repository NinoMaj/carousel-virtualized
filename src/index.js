export { default as CarouselVirtualized } from './CarouselVirtualized';
