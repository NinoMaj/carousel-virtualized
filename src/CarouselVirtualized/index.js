/* eslint-disable */
import React from 'react';
import { PropTypes } from 'prop-types';
import { FixedSizeList as Carousel } from 'react-window';
import throttle from 'lodash/throttle';

import { DIRECTION } from '../enums/direction';
import { EVENT_NAME } from '../enums/eventName';
import { KEYBOARD_EVENT } from '../enums/keyboardEvent';
import { RESIZE_THROTTLE_THRESHOLD } from '../consts';


class CarouselVirtualized extends React.PureComponent {
  static propTypes = {
    children: PropTypes.func.isRequired,
    disableDrag: PropTypes.bool,
    disableTouch: PropTypes.bool,
    height: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
    initialScrollOffset: PropTypes.number,
    innerClassName: PropTypes.bool,
    itemCount: PropTypes.number.isRequired,
    itemData: PropTypes.any,
    itemSize: PropTypes.number,
    outerClassName: PropTypes.string,
    outerStyle: PropTypes.string,
    overscanCount: PropTypes.number,
    slideCount: PropTypes.number,
    style: PropTypes.object,
    width: PropTypes.number,
  }

  static defaultProps = {
    disableDrag: false,
    disableTouch: false,
    height: null,
    initialScrollOffset: 0,
    innerClassName: null,
    outerClassName: null,
    overscanCount: 1,
    style: {},
    slideCount: null,
    width: null,
  }

  static slidesMoved(deltaX, deltaY, itemSize) {
    const threshold = 0.1;
    const bigDrag = Math.abs(Math.round(deltaX / itemSize));
    const smallDrag = (Math.abs(deltaX) >= (itemSize * threshold)) ? 1 : 0;

    return deltaX < 0
      ? Math.max(smallDrag, bigDrag)
      : -Math.max(bigDrag, smallDrag);
  }

  constructor(props) {
    super(props);
    this.state = {
      containerHeight: null,
      containerWidth: null,
      currentIndex: this.props.slideIndex,
      eventName: EVENT_NAME.INITIAL,
      isMounted: false,
      isMouseDragActive: false,
      isTouchDragActive: false,
      startX: 0,
      startY: 0,
    };

    this.moveTimer = null;

    this.containerRef = React.createRef();
    this.listRef = React.createRef();

    this.scrollToItem = this.scrollToItem.bind(this);
    this.onItemsRendered = this.onItemsRendered.bind(this);
    this.onResize = this.onResize.bind(this);

    this.handleOnKeyDown = this.handleOnKeyDown.bind(this);

    this.handleLeftArrowClick = this.handleLeftArrowClick.bind(this);
    this.handleRightArrowClick = this.handleRightArrowClick.bind(this);

    this.handleOnMouseDown = this.handleOnMouseDown.bind(this);
    this.handleOnMouseMove = this.handleOnMouseMove.bind(this);
    this.handleOnMouseClick = this.handleOnMouseClick.bind(this);

    this.handleTouchStart = this.handleTouchStart.bind(this);
    this.handleTouchMove = this.handleTouchMove.bind(this);
    this.handleTouchEnd = this.handleTouchEnd.bind(this);
    this.handleTouchCancel = this.handleTouchCancel.bind(this);
    this.itemSize = this.itemSize.bind(this);
  }

  componentDidMount() {
    this.setState({
      containerWidth: this.containerRef.current.clientWidth,
      containerHeight: this.containerRef.current.clientHeight,
      isMounted: true,
    });

    if (this.props.autofocus) {
      this.containerRef.current.focus();
    }

    window.addEventListener('resize', this.onResize);
  }


  componentDidUpdate(prevProps) {
    if (prevProps.slideIndex !== this.props.slideIndex) {
      this.setState({ currentIndex: this.props.slideIndex });
      this.scrollToItem(this.props.slideIndex);
    }
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
    window.cancelAnimationFrame.call(window, this.moveTimer);
    this.moveTimer = null;
  }

  onResize() {
    this.throttledResize();
  }

  throttledResize = throttle(() => {
    this.setState({
      containerWidth: this.containerRef.current.clientWidth,
      containerHeight: this.containerRef.current.clientHeight,
      eventName: EVENT_NAME.RESIZE,
    }, () => this.scrollToItem(this.props.slideIndex));
    // throttle time, option, leading timeout, trailing timout
  }, RESIZE_THROTTLE_THRESHOLD, {}, false, true);

  handleOnKeyDown(event) {
    event.preventDefault();
    event.stopPropagation();
    if (event.which === KEYBOARD_EVENT.LEFT_ARROW || event.which === KEYBOARD_EVENT.RIGHT_ARROW) {
      this.changeSlide(event.which === KEYBOARD_EVENT.LEFT_ARROW ? DIRECTION.LEFT : DIRECTION.RIGHT, EVENT_NAME.KEYDOWN);
    }
  }

  handleOnMouseDown(event) {
    event.stopPropagation();
    if (this.props.disableDrag) {
      return;
    }

    event.preventDefault();
    this.onDragStart(event.screenX, event.screenY, true, false);
  }

  handleOnMouseMove(event) {
    if (this.props.disableDrag || !this.state.isMouseDragActive) {
      return;
    }

    event.preventDefault();
    event.persist();

    this.onDragMove(event.screenX, event.screenY);
  }

  handleOnMouseClick() {
    if (this.props.disableDrag || !this.state.isMouseDragActive) {
      return
    };

    if (this.state.mouseIsMoving) {
      event.preventDefault();
    }

    this.onDragEnd();
  }

  handleTouchStart(event) {
    if (this.props.disableTouch) {
      return;
    }

    const touch = event.targetTouches[0];
    this.onDragStart(touch.screenX, touch.screenY, false, true);
  }

  handleTouchMove(event) {
    if (this.props.disableTouch) {
      return;
    }

    window.cancelAnimationFrame.call(window, this.moveTimer);

    const touch = event.targetTouches[0];
    this.onDragMove(touch.screenX, touch.screenY);
  }

  handleTouchEnd() {
    this.endTouchMove();
  }
  handleTouchCancel() {
    this.endTouchMove();
  }

  endTouchMove() {
    if (this.props.disableTouch) return;
    this.onDragEnd();
  }

  onDragStart(startX, startY, isMouseDragActive, isTouchDragActive) {
    window.cancelAnimationFrame.call(window, this.moveTimer);

    this.setState({
      isMouseDragActive,
      isTouchDragActive,
      startX,
      startY,
    });
  }

  onDragMove(screenX, screenY) {
    this.moveTimer = window.requestAnimationFrame.call(window, () => {
      this.setState({
        deltaX: screenX - this.state.startX,
        deltaY: screenY - this.state.startY,
        mouseIsMoving: this.state.isMouseDragActive,
      });
    });
  }

  onDragEnd() {
    window.cancelAnimationFrame.call(window, this.moveTimer);

    this.computeNextSlide();

    this.setState({
      deltaX: 0,
      deltaY: 0,
      isMouseDragActive: false,
      isTouchDragActive: false,
      mouseIsMoving: false,
    });
  }

  computeNextSlide() {
    const slidesMoved = CarouselVirtualized.slidesMoved(
      this.state.deltaX,
      this.state.deltaY,
      this.itemSize(),
    );

    this.changeSlide(slidesMoved);
  }

  changeSlide(change, eventName) {
    const { itemCount } = this.props;
    const adjustedIdx = this.state.currentIndex + change;
    let newIdx;
    if (adjustedIdx >= itemCount) {
      newIdx = 0;
    } else if (adjustedIdx < 0) {
      newIdx = itemCount - 1;
    } else {
      newIdx = adjustedIdx;
    }
    this.setState({ currentIndex: newIdx, eventName });
    this.scrollToItem(newIdx);
  }

  onItemsRendered({
    overscanStartIndex,
    overscanStopIndex,
    visibleStartIndex,
    visibleStopIndex,
  }) {
    return this.props.onItemsRendered({
      eventName: this.state.eventName,
      carouselName: this.props.carouselName,
      overscanStartIndex,
      overscanStopIndex,
      visibleStartIndex,
      visibleStopIndex,
    });
  }

  scrollToItem(index) {
    this.listRef.current.scrollToItem(index, 'center');
  }

  handleLeftArrowClick() {
    this.changeSlide(DIRECTION.LEFT, EVENT_NAME.LEFT_ARROW_CLICK);
  }

  handleRightArrowClick() {
    this.changeSlide(DIRECTION.RIGHT, EVENT_NAME.RIGHT_ARROW_CLICK);
  }

  itemSize() {
    const { itemSize, width, slideCount } = this.props;

    return itemSize
      ? itemSize
      : slideCount
        ? width / slideCount
        : this.state.containerWidth || 0;
  }

  render() {
    const {
      children,
      outerClassName,
      height,
      initialScrollOffset,
      innerClassName,
      itemCount,
      itemData,
      itemSize,
      overscanCount,
      slideCount,
      style,
      outerStyle,
      width,
    } = this.props;

    const dataForRenderer = {
      width,
      height,
      containerHeight: this.state.containerHeight,
      containerWidth: this.state.containerWidth,
      itemData,
    };

    const calculatedItemSize = itemSize
      ? itemSize
      : slideCount
        ? width / slideCount
        : this.state.containerWidth || 0;
    const calculatedHeight = height || this.state.containerHeight || 0;
    const calculatedWidth = itemSize && slideCount
      ? itemSize * slideCount
      : width || this.state.containerWidth || 0;
    const calculatedInitialScrollOffset = initialScrollOffset
      ? initialScrollOffset
      : (itemSize || this.state.containerWidth) * this.state.currentIndex;

    return (
      <React.Fragment>
        {this.props.leftArrow && this.props.leftArrow({ onClick: this.handleLeftArrowClick })}
        <div
          onClick={this.handleOnMouseClick}
          onTouchCancel={this.handleTouchCancel}
          onTouchEnd={this.handleTouchEnd}
          onTouchMove={this.handleTouchMove}
          onTouchStart={this.handleTouchStart}
          className={outerClassName}
          onKeyDown={this.handleOnKeyDown}
          onMouseDown={this.handleOnMouseDown}
          onMouseMove={this.handleOnMouseMove}
          ref={this.containerRef}
          role="listbox"
          style={{
            width: width ? 'auto' : '100%',
            height: height && typeof height === 'number' ? 'auto' : '100%',
            ...outerStyle,
          }}
          tabIndex={0}
        >
          {this.state.isMounted ? (
            <Carousel
              className={innerClassName}
              direction="horizontal"
              height={calculatedHeight}
              initialScrollOffset={calculatedInitialScrollOffset}
              itemCount={itemCount}
              itemData={dataForRenderer}
              itemSize={calculatedItemSize}
              onItemsRendered={this.onItemsRendered}
              overscanCount={overscanCount}
              ref={this.listRef}
              style={{ overflow: 'hidden', ...style }}
              width={calculatedWidth}
            >
              {children}
            </Carousel>
          ) : null}
        </div>
        {this.props.rightArrow && this.props.rightArrow({ onClick: this.handleRightArrowClick })}
      </React.Fragment>
    );
  }
}

export default CarouselVirtualized;
