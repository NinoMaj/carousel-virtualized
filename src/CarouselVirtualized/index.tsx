import * as React from "react";
import * as PropTypes from 'prop-types';
import throttle from "lodash.throttle";
import {
  FixedSizeList as Carousel,
  FixedSizeList,
  ListChildComponentProps,
} from "react-window";

import { DIRECTION } from "../enums/direction";
import { EVENT_NAME } from "../enums/eventName";
import { KEYBOARD_EVENT } from "../enums/keyboardEvent";
import { RESIZE_THROTTLE_THRESHOLD } from "../consts";

type CarouselVirtualizedProps = {
  autofocus?: boolean;
  carouselName?: string;
  children: React.ComponentType<ListChildComponentProps>;
  disableDrag?: boolean;
  disableTouch?: boolean;
  height?: string | number;
  initialScrollOffset?: number;
  innerClassName?: string;
  itemCount: number;
  itemData?: any;
  itemSize?: number;
  leftArrow?(any): any; // TODO: add type
  onItemsRendered?(any): any; // TODO: add type
  outerClassName?: string;
  outerStyle?: object;
  overscanCount?: number;
  rightArrow?(any): any; // TODO: add type
  slideCount?: number;
  slideIndex?: number;
  style?: object;
  width?: number;
};

interface CarouselVirtualizedState {
  containerHeight: number;
  containerWidth: number;
  currentIndex: number;
  deltaX: number;
  eventName: string;
  isMounted: boolean;
  isMouseDragActive: boolean;
  isTouchDragActive: boolean;
  mouseIsMoving: boolean;
  startX: number;
  startY: number;
}

class CarouselVirtualized extends React.Component<
  CarouselVirtualizedProps,
  CarouselVirtualizedState
> {
  public static propTypes = {
    autofocus: PropTypes.bool,
    carouselName: PropTypes.string,
    children: PropTypes.func.isRequired,
    disableDrag: PropTypes.bool,
    disableTouch: PropTypes.bool,
    height: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
    initialScrollOffset: PropTypes.number,
    innerClassName: PropTypes.bool,
    itemCount: PropTypes.number.isRequired,
    itemData: PropTypes.any,
    itemSize: PropTypes.number,
    leftArrow: PropTypes.func,
    rightArrow: PropTypes.func,
    onItemsRendered: PropTypes.func,
    outerClassName: PropTypes.string,
    outerStyle: PropTypes.object,
    overscanCount: PropTypes.number,
    slideCount: PropTypes.number.isRequired,
    slideIndex: PropTypes.number,
    style: PropTypes.object,
    width: PropTypes.number,
  }

  public static defaultProps = {
    autofocus: false,
    carouselName: '', 
    disableDrag: false,
    disableTouch: false,
    height: null,
    initialScrollOffset: 0,
    innerClassName: null,
    itemSize: null,
    itemData: null,
    leftArrow: () => {},
    rightArrow: () => {},
    onItemsRendered: () => {},
    outerClassName: null,
    outerStyle: {},
    overscanCount: 1,
    style: {},
    slideCount: null,
    slideIndex: 0,
    width: null,
  }

  public static slidesMoved(deltaX, itemSize) {
    const threshold = 0.1;
    const bigDrag = Math.abs(Math.round(deltaX / itemSize));
    const smallDrag = Math.abs(deltaX) >= itemSize * threshold ? 1 : 0;
    return deltaX < 0
      ? Math.max(smallDrag, bigDrag)
      : -Math.max(bigDrag, smallDrag);
  }

  state = {
    containerHeight: null,
    containerWidth: null,
    currentIndex: this.props.slideIndex,
    deltaX: 0,
    eventName: EVENT_NAME.INITIAL,
    isMounted: false,
    isMouseDragActive: false,
    isTouchDragActive: false,
    mouseIsMoving: false,
    startX: 0,
    startY: 0,
  };

  private moveTimer: number | null = null;

  private containerRef: React.RefObject<HTMLDivElement> = React.createRef();
  // TODO: add type, something like
  // private carouselRef: React.Ref<FixedSizeList> = React.createRef();
  private carouselRef: any = React.createRef();

  public componentDidMount() {
    this.setState({
      containerWidth: this.containerRef.current.clientWidth,
      containerHeight: this.containerRef.current.clientHeight,
      isMounted: true
    });

    if (this.props.autofocus) {
      this.containerRef.current.focus();
    }

    window.addEventListener("resize", this.onResize);
  }

  public componentDidUpdate(prevProps: CarouselVirtualizedProps) {
    if (prevProps.slideIndex !== this.props.slideIndex) {
      this.setState({ currentIndex: this.props.slideIndex });
      this.scrollToItem(this.props.slideIndex);
    }
  }

  public componentWillUnmount() {
    window.removeEventListener("resize", this.onResize);
    window.cancelAnimationFrame.call(window, this.moveTimer);
    this.moveTimer = null;
  }

  private onResize() {
    this.throttledResize();
  }

  private throttledResize = throttle(
    () => {
      this.setState(
        {
          containerWidth: this.containerRef.current.clientWidth,
          containerHeight: this.containerRef.current.clientHeight,
          eventName: EVENT_NAME.RESIZE,
        },
        () => this.scrollToItem(this.props.slideIndex)
      );
    },
    RESIZE_THROTTLE_THRESHOLD,
    {
      leading: false,
      trailing: true
    },
  );

  private handleOnKeyDown: React.KeyboardEventHandler = (
    event: React.KeyboardEvent
  ) => {
    event.preventDefault();
    event.stopPropagation();
    if (
      event.which === KEYBOARD_EVENT.LEFT_ARROW ||
      event.which === KEYBOARD_EVENT.RIGHT_ARROW
    ) {
      this.changeSlide(
        event.which === KEYBOARD_EVENT.LEFT_ARROW
          ? DIRECTION.LEFT
          : DIRECTION.RIGHT,
        EVENT_NAME.KEYDOWN,
      );
    }
  };

  private handleOnMouseDown: React.MouseEventHandler = (event: React.MouseEvent) => {
    event.stopPropagation();
    if (this.props.disableDrag) {
      return;
    }
    event.preventDefault();
    this.onDragStart(event.screenX, true, false);
  };

  private handleOnMouseMove: React.MouseEventHandler = (event: React.MouseEvent) => {
    if (this.props.disableDrag || !this.state.isMouseDragActive) {
      return;
    }

    event.preventDefault();
    event.persist();
    this.onDragMove(event.screenX);
  };

  private handleOnMouseClick: React.MouseEventHandler = () => {
    if (this.props.disableDrag || !this.state.isMouseDragActive) {
      return;
    }

    if (this.state.mouseIsMoving) {
      event.preventDefault();
    }

    this.onDragEnd();
  };

  private handleTouchStart: React.TouchEventHandler = (event: React.TouchEvent) => {
    if (this.props.disableTouch) {
      return;
    }
    const touch = event.targetTouches[0];
    this.onDragStart(touch.screenX, false, true);
  }

  private handleTouchMove: React.TouchEventHandler = (event: React.TouchEvent) => {
    if (this.props.disableTouch) {
      return;
    }
    window.cancelAnimationFrame.call(window, this.moveTimer);
    const touch = event.targetTouches[0];
    this.onDragMove(touch.screenX);
  }

  private handleTouchEnd: React.TouchEventHandler = () => {
    this.endTouchMove();
  }

  private handleTouchCancel: React.TouchEventHandler = () => {
    this.endTouchMove();
  }

  private endTouchMove() {
    if (this.props.disableTouch) {
      return;
    }
    this.onDragEnd();
  }

  private onDragStart(startX: number, isMouseDragActive: boolean, isTouchDragActive: boolean) {
    window.cancelAnimationFrame.call(window, this.moveTimer);
    this.setState({
      isMouseDragActive,
      isTouchDragActive,
      startX,
    });
  }

  private onDragMove(screenX: number) {
    this.moveTimer = window.requestAnimationFrame.call(window, () => {
      this.setState(state => ({
        deltaX: screenX - state.startX,
        mouseIsMoving: state.isMouseDragActive,
      }));
    });
  }

  private onDragEnd() {
    window.cancelAnimationFrame.call(window, this.moveTimer);
    this.computeNextSlide();
    this.setState({
      deltaX: 0,
      isMouseDragActive: false,
      isTouchDragActive: false,
      mouseIsMoving: false
    });
  }

  private computeNextSlide() {
    const slidesMoved = CarouselVirtualized.slidesMoved(
      this.state.deltaX,
      this.itemSize()
    );
    this.changeSlide(slidesMoved, this.state.eventName);
  }

  private changeSlide(change: number, eventName: string) {
    const { itemCount } = this.props;
    const adjustedIdx = this.state.currentIndex + change;
    let newIdx;
    if (adjustedIdx >= itemCount) {
      newIdx = 0;
    } else if (adjustedIdx < 0) {
      newIdx = itemCount - 1;
    } else {
      newIdx = adjustedIdx;
    }
    this.setState({ currentIndex: newIdx, eventName });
    this.scrollToItem(newIdx);
  }

  private onItemsRendered({
    overscanStartIndex,
    overscanStopIndex,
    visibleStartIndex,
    visibleStopIndex,
  }) {
    return this.props.onItemsRendered({
      eventName: this.state.eventName,
      carouselName: this.props.carouselName,
      overscanStartIndex,
      overscanStopIndex,
      visibleStartIndex,
      visibleStopIndex,
    });
  }

  private scrollToItem(index: number) {
    this.carouselRef.current.scrollToItem(index, "center");
  }

  private handleLeftArrowClick() {
    this.changeSlide(DIRECTION.LEFT, EVENT_NAME.LEFT_ARROW_CLICK);
  }

  private handleRightArrowClick() {
    this.changeSlide(DIRECTION.RIGHT, EVENT_NAME.RIGHT_ARROW_CLICK);
  }

  private itemSize() {
    const { itemSize, width, slideCount } = this.props;
    return itemSize
      ? itemSize
      : slideCount
        ? width / slideCount
        : this.state.containerWidth || 0;
  }

  public render() {
    const {
      children,
      outerClassName,
      height,
      initialScrollOffset,
      innerClassName,
      itemCount,
      itemData,
      itemSize,
      overscanCount,
      slideCount,
      style,
      outerStyle,
      width
    } = this.props;
    const dataForRenderer = {
      width,
      height,
      containerHeight: this.state.containerHeight,
      containerWidth: this.state.containerWidth,
      itemData,
    };
    const calculatedItemSize: number = itemSize
      ? itemSize
      : slideCount
        ? width / slideCount
        : this.state.containerWidth || 0;
    const calculatedHeight: number|string = height || this.state.containerHeight || 0;
    const calculatedWidth: number =
      itemSize && slideCount
        ? itemSize * slideCount
        : width || this.state.containerWidth || 0;
    const calculatedInitialScrollOffset: number = initialScrollOffset
      ? initialScrollOffset
      : (itemSize || this.state.containerWidth) * this.state.currentIndex;

    return (
      <React.Fragment>
        {this.props.leftArrow &&
          this.props.leftArrow({ onClick: this.handleLeftArrowClick })}
        <div
          className={outerClassName}
          onClick={this.handleOnMouseClick}
          onKeyDown={this.handleOnKeyDown}
          onMouseDown={this.handleOnMouseDown}
          onMouseMove={this.handleOnMouseMove}
          onTouchCancel={this.handleTouchCancel}
          onTouchEnd={this.handleTouchEnd}
          onTouchMove={this.handleTouchMove}
          onTouchStart={this.handleTouchStart}
          ref={this.containerRef}
          role="listbox"
          style={{
            width: width ? "auto" : "100%",
            height: height && typeof height === "number" ? "auto" : "100%",
            ...outerStyle
          }}
        >
          {this.state.isMounted ? (
            <Carousel
              className={innerClassName}
              direction="horizontal"
              height={calculatedHeight}
              initialScrollOffset={calculatedInitialScrollOffset}
              itemCount={itemCount}
              itemData={dataForRenderer}
              itemSize={calculatedItemSize}
              onItemsRendered={this.onItemsRendered}
              overscanCount={overscanCount}
              ref={this.carouselRef}
              style={{ overflow: "hidden", ...style }}
              width={calculatedWidth}
            >
              {children}
            </Carousel>
          ) : null}
        </div>
        {this.props.rightArrow &&
          this.props.rightArrow({ onClick: this.handleRightArrowClick })}
      </React.Fragment>
    );
  }
}
export default CarouselVirtualized;
