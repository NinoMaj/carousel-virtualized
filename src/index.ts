import { Carousel } from './components/Carousel/Carousel';
import { CarouselVirtualized } from './components/CarouselVirtualized/CarouselVirtualized';
import { Slider } from './components/Slider/Slider';
import { Alignment } from './enums/Alignment';
import { ArrowStep } from './enums/ArrowStep';
import { Direction } from './enums/Direction';
import { EventName } from './enums/EventName';
import { KeyboardButton } from './enums/KeyboardButton';

export {
  Alignment,
  ArrowStep,
  Carousel,
  Direction,
  EventName,
  Slider,
  KeyboardButton,
  CarouselVirtualized,
};
