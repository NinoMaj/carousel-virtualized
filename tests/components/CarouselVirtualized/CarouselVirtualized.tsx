import * as React from 'react';
import { fireEvent, render } from 'react-testing-library';
import 'react-testing-library/cleanup-after-each';

import { CarouselVirtualized } from '../../../src';

describe('CarouselVirtualized', () => {
  let slideRenderer;
  let defaultProps;
  let onEvent;
  let onItemsRendered;

  class Slide extends React.PureComponent {
    public render() {
      return slideRenderer(this.props);
    }
  }

  beforeEach(() => {
    jest.useFakeTimers();

    onEvent = jest.fn();
    onItemsRendered = jest.fn();

    slideRenderer = jest.fn(({ style, index, ...rest }) => (
      <div style={style} data-testid={`slide-${index}`}>{`slide-${index}`}</div>
    ));

    defaultProps = {
      children: Slide,
      itemCount: 100,
      onEvent,
      onItemsRendered,
    };
  });

  it('should render an empty list', () => {
    render(
      <CarouselVirtualized
        {...defaultProps}
        itemCount={0}
        height={400}
        width={800}
        itemSize={400}
      />,
    );
    expect(slideRenderer).not.toHaveBeenCalled();
    expect(onItemsRendered).not.toHaveBeenCalled();
  });

  it('should render slides', () => {
    render(
      <CarouselVirtualized
        {...defaultProps}
        height={400}
        width={800}
        itemSize={400}
      />,
    );
    expect(slideRenderer).toHaveBeenCalledTimes(4);
    expect(onItemsRendered.mock.calls).toMatchSnapshot();
  });

  it('if height is not provided should have height of it\'s container', () => {
    const { getByTestId } = render(
      <div style={{ width: '800px', height: '400px' }}>
        <CarouselVirtualized
          {...defaultProps}
          width={800}
          itemSize={25}
        />
      </div>,
    );
    const container = getByTestId('carouselVirtualizedContainer');

    expect(container.style.height).toBe('100%');
  });

  it('if width is not provided should have width of it\'s container', () => {
    const { getByTestId } = render(
      <div style={{ width: '800px', height: '400px' }}>
        <CarouselVirtualized
          {...defaultProps}
          height={400}
          itemSize={25}
        />
      </div>,
    );
    const container = getByTestId('carouselVirtualizedContainer');

    expect(container.style.width).toBe('100%');
    });

  it('should render current index', () => {
    const { getByTestId } = render(
      <CarouselVirtualized
        {...defaultProps}
        currentIndex={5}
        height={400}
        width={800}
        itemSize={800}
      />,
    );
    const container = getByTestId('carouselVirtualizedContainer');
    const secondSlideInDom = container.children[0].children[0].children[1];

    expect(secondSlideInDom.innerHTML).toBe('slide-5');
    expect(onItemsRendered.mock.calls).toMatchSnapshot();
  });

  it('should fire onEvent with new index when arrow is clicked', () => {
    const rightArrow = ({ onClick }) => <button onClick={onClick}>RightArrow</button>;
    const { getByText } = render(
      <CarouselVirtualized
        {...defaultProps}
        currentIndex={5}
        height={400}
        width={800}
        itemSize={800}
        rightArrow={rightArrow}
      />,
    );
    expect(onEvent).not.toHaveBeenCalled();
    fireEvent.click(getByText('RightArrow'));
    expect(onEvent).toHaveBeenCalled();
    expect(onEvent.mock.calls[0][0].newIndex).toBe(6);
    expect(onEvent.mock.calls).toMatchSnapshot();
  });
});
