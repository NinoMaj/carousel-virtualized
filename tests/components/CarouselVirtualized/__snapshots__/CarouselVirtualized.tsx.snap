// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselVirtualized should fire onEvent with new index when arrow is clicked 1`] = `
Array [
  Array [
    Object {
      "eventName": "rightarrowclick",
      "newIndex": 6,
    },
  ],
]
`;

exports[`CarouselVirtualized should render current index 1`] = `
Array [
  Array [
    Object {
      "carouselName": "",
      "eventName": "initial",
      "overscanStartIndex": 4,
      "overscanStopIndex": 7,
      "visibleStartIndex": 5,
      "visibleStopIndex": 6,
    },
  ],
]
`;

exports[`CarouselVirtualized should render slides 1`] = `
Array [
  Array [
    Object {
      "carouselName": "",
      "eventName": "initial",
      "overscanStartIndex": 0,
      "overscanStopIndex": 3,
      "visibleStartIndex": 0,
      "visibleStopIndex": 2,
    },
  ],
]
`;
